respy/conditional_draws.py

```diff
@@ -162,14 +196,116 @@ def kalman_update(
+    updated_chols[-1] = shocks_cholesky
+    return updated_chols
 
-        # Update the state.
-        measurement /= sigma
-        for f in range(nfac):
-            state[f] = extended_cholcov_t[0, f + 1] * measurement
-    else:
-        log_prob_wage[0] = 0
+
+def update_cholcov(shocks_cholesky, n_wages):
+    """Calculate cholesky factors of conditional covs for all possible cases.
+
+    Parameters
+    ----------
+    shocks_cholesky : numpy.ndarray
+        cholesky factor of the covariance matrix before updating. Has
+        dimension (n_choices, n_choices)
+    n_wages : int
+        Number of wage sectors.
+
+    Returns
+    -------
+    updated_chols : numpy.ndarray
+        Array of (shape n_wages + 1, n_choices, n_choices) with the cholesky factors
+        of the updated covariance matrices for each possible observed shock. The
+        last element corresponds to not observing any shock.
+
+    """
+    n_choices = len(shocks_cholesky)
+    cov = shocks_cholesky @ shocks_cholesky.T
+
+    updated_chols = np.zeros((n_wages + 1, n_choices, n_choices))
+
+    for choice in range(n_choices):
+        reduced_cov = np.delete(np.delete(cov, choice, axis=1), choice, axis=0)
+        choice_var = cov[choice, choice]
+
+        f = np.delete(cov[choice], choice)
+
+        updated_reduced_cov = reduced_cov - np.outer(f, f) / choice_var
+        updated_reduced_chol = robust_cholesky(updated_reduced_cov)
+
+        updated_chols[choice, :choice, :choice] = updated_reduced_chol[:choice, :choice]
+        updated_chols[choice, :choice, choice + 1 :] = updated_reduced_chol[
+            :choice, choice:
+        ]
+        updated_chols[choice, choice + 1 :, :choice] = updated_reduced_chol[
+            choice:, :choice
+        ]
+        updated_chols[choice, choice + 1 :, choice + 1 :] = updated_reduced_chol[
+            choice:, choice:
+        ]
+
+    updated_chols[-1] = shocks_cholesky
+
+    return updated_chols
+
+
+@guvectorize(
+    ["f8[:, :], f8[:], f8[:, :, :], u2, f8, f8[:, :]"],
+    "(n_draws, n_choices), (n_choices), (n_wages_plus_one, n_choices, n_choices), (), "
+    "() -> (n_draws, n_choices)",
+    nopython=True,
+)
+def calculate_conditional_draws2(
+    base_draws, updated_mean, updated_chols, chol_index, huge_float, conditional_draw
+):
+    """Calculate the conditional draws from base draws, updated means and updated chols.
+
+    Parameters
+    ----------
+    base_draws : np.ndarray
+        iid standard normal draws
+    updated_mean : np.ndarray
+        conditional mean, given the observed shock. Contains the observed shock in the
+        corresponding position.
+    updated_chols : np.ndarray
+        cholesky factor of conditional covariance, given the observed shock. If there is
+        no measurement error, it contains a zero column and row at the position of the
+        observed shock.
+    chol_index : float
+        index of the relevant updated cholesky factor
+    huge_float : float
```

Comment:
Why an input? Normally, we import and use ``HUGE_FLOAT`` straight away.

Reply:
Numba complained that it can't determine the type of `HUGE_FLOAT`.

Reply:
Sounds like a typical missing import. We use it in ``njit`` functions, so I was sure it also works for ``guvectorize``.